include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,413 @@
+
+    const ::std::size_t __n = __in_rng.size();
+
+    if (__n == 0)
+        return sycl::event{};
+
+    static_assert(_Inclusive, "Single-pass scan only available for inclusive scan");
+    static_assert(oneapi::dpl::unseq_backend::__has_known_identity<_BinaryOp, _Type>::value,
+                  "Only binary operators with known identity values are supported");
+
+    assert("This device does not support 64-bit atomics" &&
+           (sizeof(_Type) < 64 || __queue.get_device().has(sycl::aspect::atomic64)));
+
+    // Next power of 2 greater than or equal to __n
+    auto __n_uniform = __n;
+    if ((__n_uniform & (__n_uniform - 1)) != 0)
+        __n_uniform = oneapi::dpl::__internal::__dpl_bit_floor(__n) << 1;
+
+    // Perform a single-work group scan if the input is small
+    if (oneapi::dpl::__par_backend_hetero::__group_scan_fits_in_slm<_Type>(__queue, __n, __n_uniform))
+    {
+        return oneapi::dpl::__par_backend_hetero::__parallel_transform_scan_single_group(
+            oneapi::dpl::execution::__dpl::make_device_policy(__queue), ::std::forward<_InRange>(__in_rng),
+            ::std::forward<_OutRange>(__out_rng), __n, oneapi::dpl::__internal::__no_op{}, unseq_backend::__no_init_value<_Type>{}, __binary_op, ::std::true_type{});
+    }
+
+    constexpr ::std::size_t __workgroup_size = _KernelParam::workgroup_size;
+    constexpr ::std::size_t __data_per_workitem = _KernelParam::data_per_workitem;
+
+    // Avoid non_uniform n by padding up to a multiple of workgroup_size
+    std::uint32_t __elems_in_tile = __workgroup_size * __data_per_workitem;
+    ::std::size_t __num_wgs = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __elems_in_tile);
+
+    constexpr int __status_flag_padding = SUBGROUP_SIZE;
+    std::uint32_t __status_flags_size = __num_wgs + 1 + __status_flag_padding;
+
+    _FlagStorageType* __status_flags = sycl::malloc_device<_FlagStorageType>(__status_flags_size, __queue);
+    _Type* __status_vals_full = sycl::malloc_device<_Type>(__status_flags_size, __queue);
+    _Type* __status_vals_partial = sycl::malloc_device<_Type>(__status_flags_size, __queue);
+
+    assert(__status_flags && __status_vals_full && __status_vals_partial);
+
+    auto __fill_event = __lookback_init_submitter<_FlagType, _Type, _BinaryOp, _LookbackInitKernel>{}(
+        __queue, __status_flags, __status_vals_partial, __status_flags_size, __status_flag_padding);
+
+    ::std::size_t __current_num_wgs = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __elems_in_tile);
+    ::std::size_t __current_num_items = __current_num_wgs * __workgroup_size;
+
+    auto __prev_event =
+        __lookback_submitter<__data_per_workitem, __workgroup_size, _Type, _FlagType, _LookbackKernel>{}(
+            __queue, __fill_event, __in_rng, __out_rng, __binary_op, __n, __status_flags, __status_flags_size,
+            __status_vals_full, __status_vals_partial, __current_num_items);
+
+    if (0)
```

Comment:
Are we planning to keep this synchronous or add back in the host task submission to support asynchronous calls?

Reply:
That is a good question. I do not like how it is written currently, as it introduces a synchronization while still giving the appearance that the algorithm itself is asynchronous. However, the asynchronous free introduces an unacceptably high slowdown, which I suspect is potentially a compiler bug.

I am not sure what the right call would be here, but in order to actually see a speedup with the kernel template, the synchronous free is needed currently unfortunately.

Reply:
I think the best course of action would be to document this synchronous for now and bring it up with the compiler team to be improved.

Reply:
This makes sense to me as well unless our current targets require asynchronous versions. (I don't think they do)